### ISSUE TITLE
Remove uses of opts::get() from as much code as possible

### ISSUE THREAD

**Issue by jdm** [servo/servo#22854]
Right now the code in components/config has a bunch of code depending on it because the command line arguments for Servo are global singleton that any code can access. This makes changing the command line code for Servo frustrating, because a lot of code gets rebuilt unnecessarily. We should remove as many uses of opts::get() as possible, and replace them with values that are passed into appropriate constructors via the code in components/servo/lib.rs.

We can focus on one component at a time:
* [ ] bluetooth
* [ ] compositing
* [ ] constellation
* [ ] layout
* [ ] layout_thread
* [ ] metrics
* [ ] net
* [ ] profile
* [ ] profile_traits
* [ ] script
* [ ] style

**Comment by JacksonCoder** [comment]
@highfive: assign me

**Comment by peterjoel** [comment]
Couldn't `opts` just be a non-persistent `prefs` override?

**Comment by jdm** [comment]
I'm not sure I understand what you're suggesting. Do you have an example?

**Comment by peterjoel** [comment]
> I'm not sure I understand what you're suggesting. Do you have an example?

I am suggesting that all CLI arguments would actually set a preference, which would not be persisted. Instead of accessing `opts`, any configurable code would just access `prefs`. The actual value would be evaluated in ascending order of precedence: `prefs.json`, user preferences, CLI argument.

**Comment by jdm** [comment]
Ah, I see. The advantage of keeping the command line arguments separate from preferences is that the arguments are not mutable. This is one reason why it's safe to remove access to the global arguments storage and pass the values directly to the code that uses them, allowing us to remove dependencies.

**Comment by peterjoel** [comment]
@jdm It's not ideal then that _some_ opts are actually duplicated as (mutable) prefs, so you _could_ get a different value for `layout.threads`, depending on if you access it via `opts` or `prefs`.

**Comment by jdm** [comment]
That's true! Anything that is duplicated should be removed from the immutable options structure and any users changed to use the preference.

**Comment by jdm** [comment]
With the caveat that using the command line option should mutate the preference.

**Comment by peterjoel** [comment]
Rather than replace all instances of `opts` with dependency injection, another approach is to split opts into two levels: globals (which should change infrequently) and crate-level options (which can develop alongside the features of the crate itself). For example, `bluetooth` only uses one option: `headless`, which is global. `compositing` uses some globals (`exit_after_load` maybe?), but also more specific options like `device_pixels_per_px`.

Each crate would define its own options and provide access to both its own and the global options. 

If you add or remove a global option then all the component crates will still need to be rebuilt, but adding a compositing option won't affect bluetooth.

**Comment by jdm** [comment]
You say headless is global, but it's only used by 3 crates (or 5 modules). I guess I don't see a big benefit to continuing to expose a set of global options, given how restricted their use generally is.

**Comment by peterjoel** [comment]
Ok, that's fair.

**Comment by JacksonCoder** [comment]
@jdm I've finally gotten around to getting this fixed. Just to be clear, should the goal be to pass the arguments we're using inside different modules to their constructors, and then storing it within that struct? Or is there a more idiomatic way to do it?

**Comment by jdm** [comment]
Yes, store the values in places where they can be accessed when they need to be, or take them as arguments for any code that uses them at initialization only.

**Comment by oneturkmen** [comment]
@highfive: assign me

**Comment by oneturkmen** [comment]
@jdm just fyi, there is no `opts::get()` (neither `headless`) in the `bluetooth` module. Working on `compositing` now.

**Comment by gootorov** [comment]
@jdm, @oneturkmen, looks like it got removed in https://github.com/servo/servo/commit/9b3d76b82d9a4365538a1f01d71ca77925c5f768.

**Comment by oneturkmen** [comment]
@jdm, wanted to ask for an advice. I am done with `compositing` - should I continue with the rest of the components and commit under a single PR, or rather have PR per component?

**Comment by jdm** [comment]
One PR each would be fine.

**Comment by oneturkmen** [comment]
@jdm take a look at PR #23493 and let me know what you think.

**Comment by gootorov** [comment]
There is more code code that can be removed that uses `opts::get()` in `ports/glutin`. The initialization points are in `ports/glutin/app.rs` and `ports/glutin/main2.rs`.

**Comment by jdm** [comment]
I am unconcerned about the code in ports/, since those are application crates and there is not code that can pass the options to it.

**Comment by oneturkmen** [comment]
@jdm changes in `compositing` have been merged (thank you for the help!). Changes to `constellation` are out as well (#23506). Proceeding with the `layout` and `layout_thread` components.

**Comment by oneturkmen** [comment]
Btw, `servo/lib.rs` does not seem to depend on `layout` ... or does it? (i.e. I see no way of replacing `opts::get()` via `servo/lib.rs`).

**Comment by jdm** [comment]
We can get options to that code transitively through the layout threads, which are created in components/constellation by the pipeline code.

**Comment by oneturkmen** [comment]
@jdm, I am a bit confused by components/metrics. So, there is one (non-pub) function which uses opts::get, namely `set_metrics`. It is used by PaintTimeMetrics and InteractiveMetrics type implementations, which are used in components/layout_threads and components/dom, respectively. Should I leave opts::get() in case of PaintTimeMetrics for later? (after modifying the rest of crates)

**Comment by jdm** [comment]
Sure, that sounds reasonable.

**Comment by oneturkmen** [comment]
hey @jdm, I am currently on component/net. So, to get rid of opts::get(), I added `certificate_path: Option<String>` field in `ResourceChannelManager`. I seem to have the following compilation error:

```
error[E0382]: use of moved value: `certificate_path`
   --> components/net/resource_thread.rs:102:17
    |
96  |                 certificate_path,
    |                 ---------------- value moved here
...
102 |                 certificate_path,
    |                 ^^^^^^^^^^^^^^^^ value used here after move
    |
    = note: move occurs because `certificate_path` has type `std::option::Option<std::string::String>`, which does not implement the `Copy` trait

error: aborting due to previous error

For more information about this error, try `rustc --explain E0382`.
error: Could not compile `net`.
```

And the code where it is used:
```
let resource_manager = CoreResourceManager::new(
    user_agent,
    devtools_chan,
    time_profiler_chan,
    embedder_proxy,
    certificate_path,
);

let mut channel_manager = ResourceChannelManager {
    resource_manager,
    config_dir,
    certificate_path,
};

mem_profiler_chan.run_with_memory_reporting(
    || (channel_manager.start(public_setup_port, private_setup_port, report_port)),
    String::from("network-cache-reporter"),
    report_chan,
    |report_chan| report_chan,
);
```

I understand the error (transfer of ownership), but I am not sure how to solve it. I suspect it has to do something with `.as_ref()` or `.as_mut()`...

**Comment by jdm** [comment]
If there are multiple consumers of a non-copyable value that need to store it, the first one can use `clone()` to duplicate the vlaue. Otherwise if it's only being used inside a method but not stored, the method can accept an `Option<&str>` and it can be passed with `optional_string.as_ref().map(|s| &**s)`.

**Comment by oneturkmen** [comment]
Do you think we should get rid of opts::get() in `components/style`? There are 3 of them, but it may be redundant since these ones are part of conditional compilation, [here](https://github.com/servo/servo/blob/294d701fdaad61d571109712acf181b54e8f811b/components/style/traversal.rs#L53) and [here](https://github.com/servo/servo/blob/master/components/style/context.rs#L112).

PS. I think components [`net`](https://github.com/servo/servo/pull/23520), [`profile`](https://github.com/servo/servo/pull/23521), and [`script`](https://github.com/servo/servo/pull/23539) are done. Meanwhile, I will take a look at components/profile_traits.

**Comment by jdm** [comment]
Yes, we should remove the from the style crate.

**Comment by oneturkmen** [comment]
@jdm btw I don't think we can remove [this couple](https://github.com/servo/servo/blob/master/components/style/context.rs#L112) of opts::get() from `components/style` since the implementation of Default trait requires `fn() -> Self` function (i.e. I cannot pass `&self`).

**Comment by jdm** [comment]
Bleah. I agree. We will need to think harder about how we can remove those in the future.

**Comment by kitlith** [comment]
would some sort of global/static make sense to get rid of those? i.e. have a couple of bools, default_disable_style_sharing_cache and default_dump_style_statistics, behind a static Mutex (or a type from a lazy initialization crate) which are set shortly after reading the options?

**Comment by jdm** [comment]
That's a good idea! It would be an improvement over the status quo because the use of opts::get would stop, and it would be a global that's private to a particular crate.

**Comment by JMarlin** [comment]
@highfive: assign me
